grayscale.py

```diff
@@ -1,336 +1,1055 @@
+        self.frameRate = newFrameRate
+        if newFrameRate != 0:
+            self.frameTimeMs = 1000 // newFrameRate
 
     @micropython.native
     def update(self):
-        pass
+        self.show()
+        if self.frameRate > 0:
+            frameTimeMs = self.frameTimeMs
+            lastUpdateEnd = self.lastUpdateEnd
+            frameTimeRemaining = frameTimeMs - utime.ticks_diff(utime.ticks_ms(), self.lastUpdateEnd)
+            while frameTimeRemaining > 1:
+                buttonA.update()
+                buttonB.update()
+                buttonU.update()
+                buttonD.update()
+                buttonL.update()
+                buttonR.update()
+                utime.sleep_ms(1)
+                frameTimeRemaining = frameTimeMs - utime.ticks_diff(utime.ticks_ms(), self.lastUpdateEnd)
+            while frameTimeRemaining > 0:
+                frameTimeRemaining = frameTimeMs - utime.ticks_diff(utime.ticks_ms(), self.lastUpdateEnd)
+        self.lastUpdateEnd = utime.ticks_ms()
+
+
+    def brightness(self, c):
+        if c < 0:
+            c = 0
+        elif c > 127:
+            c = 127
+        self._state[Grayscale_StateIndex_ContrastChng] = c
+
+    def brightness_sync(self, c):
+        if c >= 128:
+            return
+        self._state[Grayscale_StateIndex_ContrastChng] = c
+        if self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Running:
+            return
+        while self._state[Grayscale_StateIndex_ContrastChng] != 0xff:
+            idle()
 
     @micropython.native
-    def brightness(self, setting):
-        self.gsGraphics1.brightness(setting)
+    def _copy_buffers(self):
+        b1 = self.buffer1 ; b2 = self.buffer2
+        _b1 = self._buffer1 ; _b2 = self._buffer2 ; _b3 = self._buffer3
+        i:int = 0
+        bs:int = self.buffer_size
+        while i < bs:
+            v1 = b1[i]
+            v2 = b2[i]
+            _b1[i] = v1 | v2
+            _b2[i] = v2
+            _b3[i] = v1 & v2
+            i += 1
+        self._state[Grayscale_StateIndex_CopyBuffs] = 0
+
+
+    if is_v1_19_1_plus:
+
+        # without gc.collect()
+        @micropython.native
+        def _display_thread_frame(self, fn:int, post_frame_adj:ptr8, fb:ptr8):
+            t0:int = utime.ticks_us()
+            spi_write = self.spi.write
+            dc = self.dc
+            dc(0)
+            spi_write(self.pre_frame_cmds)
+            dc(1)
+            spi_write(fb)
+            dc(0)
+            spi_write(post_frame_adj)
+            utime.sleep_us(int(Grayscale_pre_frame_time_us) - int(utime.ticks_diff(utime.ticks_us(), t0)) - 12)
+            t0:int = utime.ticks_us()
+            spi_write(self.post_frame_cmds)
+            spi_write(post_frame_adj)
+            state:ptr32 = self._state
+            if (fn == 2) and (_state[Grayscale_StateIndex_CopyBuffs] != 0):
+                self._copy_buffers()
+            elif state[Grayscale_StateIndex_ContrastChng] != 0xff:
```

Comment:
The gc.collect() version checks if `fn == 2` here. Don't we have to do that in this version too?

Reply:
Yeah, there were a couple of other missing changes in this version. Although it now looks like this version isn't needed with the new timings.